design/design-doc-composition.md

```diff
@@ -0,0 +1,816 @@
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromFieldPath: "spec.region"
+      toFieldPath: "spec.forProvider.location"
+      transforms:
+      - type: map
+        map:
+          us-west: "West US"
+          us-east: "East US"
+  - base:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+      spec:
+        forProvider:
+          administratorLogin: myadmin
+          resourceGroupNameSelector:
+            matchComposite: true
+          location: West US
+          sslEnforcement: Disabled
+          version: "5.6"
+          sku:
+            tier: Basic
+            capacity: 1
+            family: Gen5
+          storageProfile:
+            storageMB: 20480
+        writeConnectionSecretToRef:
+          namespace: crossplane-system
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromFieldPath: "metadata.uid"
+      toFieldPath: "spec.writeConnectionSecretToRef.name"
+    - fromFieldPath: "spec.engineVersion"
+      toFieldPath: "spec.forProvider.version"
+    - fromFieldPath: "spec.storageGB"
+      toFieldPath: "spec.forProvider.storageMB"
+      transforms:
+      - type: math
+        math:
+          multiply: 1024
+    - fromFieldPath: "spec.region"
+      toFieldPath: "spec.forProvider.location"
+      transforms:
+      - type: map
+        map:
+          us-west: "West US"
+          us-east: "East US"
+    # Specifies the (potentially sensitive) connection details that this 'to'
+    # resource should expose to the 'from' resource. Names are unique across all
+    # 'to' resources within this composition. Ignored by application resources.
+    connectionDetails:
+    - name: username
+      fromConnectionSecretKey: username
+    - name: password
+      fromConnectionSecretKey: password
+    - name: endpoint
+      fromConnectionSecretKey: endpoint
+  - base:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+      spec:
+        name: my-cool-vnet-rule
+        serverNameSelector:
+          matchComposite: true
+        resourceGroupNameSelector:
+          matchComposite: true
+        properties:
+          virtualNetworkSubnetIdRef:
+            name: sample-subnet
+        reclaimPolicy: Delete
+        providerRef:
+          name: azure-provider
+```
+
+This document proposes compositions be immutable. This property would allow
+updates to a 'from' resource to be propagated to the various 'to' resources but
+keep the set of 'to' resources stable over the lifetime of the 'from' resource,
+allowing controllers to avoid complex update and garbage collection logic.
+
+### ApplicationDefinition
+
+An `ApplicationDefinition` defines a new kind of Kubernetes custom resource that
+represents an application - for example a `Wordpress` custom resource. An
+application resource is namespaced, and may only compose resources within its
```

Comment:
I'd like to avoid assuming app devs will deploy things at the cluster scope (and generally outside of their namespace). I suspect many use cases in which an "app" consists of cluster scoped resources are actually an app trying to deploy _infrastructure_ that supports their app. Perhaps we should try to write up some example use cases in which deploying an app would affect something outside the apps namespace?

Reply:
Well, Crossplane could be considered as an app and it deploys `ClusterRole` and `ClusterRoleBinding` resources, which are cluster-scoped. There could even be cases where a few `ApplicationDefinition`s are bundled in an app to deploy different apps etc.

My point is that *app* can include anything because individuals package them whereas we know that infra managed resources are cluster-scoped because we implement them.